### ISSUE TITLE
container/list: zero List value breaks after reset

### ISSUE THREAD

**Issue by gobwas** [golang/go#39014]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.14.2 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
</pre></details>

### What did you do?
```go
var l list.List
el := l.PushBack(42)
l = list.List{} // Reset the list.
l.Remove(el)
fmt.Println(l.Len())
```

### What did you expect to see?
```bash
0
```

### What did you see instead?
```bash
-1
```

**Comment by lsytj0413** [comment]
Because the another assign **l = list.List{}** Never change the address of **l**, you can print the memory address of l with:

```
	fmt.Printf("%p\n", &l)
	el := l.PushBack(42)
	l = list.List{} // Reset the list.
	fmt.Printf("%p\n", &l)
```

**Comment by lsytj0413** [comment]
Maybe **list.List** should disable copy with https://github.com/golang/go/blob/c2eba53e7f80df21d51285879d51ab81bcfbf6bc/src/sync/cond.go#L94

**Comment by gobwas** [comment]
@lsytj0413 Yes, I know why it happens, thank you :) Anyway, the comment about `List` tells that `The zero value for List is an empty list ready to use.`.

**Comment by lsytj0413** [comment]
> @lsytj0413 Yes, I know why it happens, thank you :) Anyway, the comment about `List` tells that `The zero value for List is an empty list ready to use.`.

It's Zero Value when you reassign it，but after you called **Remove** it's length change to -1(removed one element from an empty list).

**Comment by cagedmantis** [comment]
I'm confirming that the code above reproduces. Resetting the list does not change the pointer to the list. The element still has the pointer to the original list. The Remove() function checks if the pointer to the list is the same as the pointer to list in the element.

**Comment by josharian** [comment]
> Resetting the list does not change the pointer to the list.

Seems like that could also cause memory leaks.

**Comment by adonovan** [comment]
See https://github.com/golang/go/issues/70022#issuecomment-2435981437 for possible mitigations, including merely documenting it.

**Comment by zigo101** [comment]
Similar case, which will only not panic when N <= 1 or Go version < 1.22;

```go
package main

import "container/list"

const N = 2

func CreateList() *list.List {
	for l, k := (list.List{}), 0; ; k++ {
		if l.Len() < N {
			l.PushBack(k)
		} else {
			return &l
		}
	}
}

func main() {
	var l = CreateList()
	println("====", l.Len())
	for e := l.Front(); e != nil; e = e.Next() {
		println(e.Value.(int))
	}
}
```
